Add Tankerkoenig integration

guillempages opened the pull request:
## Description:
Add a new integration for tankerkoenig.de gas station prices

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#11137

## Example entry for `configuration.yaml` (if applicable):
```yaml
tankerkoenig:
    api_key: YOUR_API_KEY
    radius: 1
    fuel_types: 
      - "diesel"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`.
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

guillempages committed:
Initial version

Parse configuration, but return a fixed value

guillempages committed:
Add basic functionality

Request real data from the server
Currently the prices are not getting updated, but the petrol station
data is real

guillempages committed:
Update values regularly

The tankerkoenig values get updated regularly with real data.

guillempages committed:
Move base functionality for the sensor to a base class

And move that to an own file, so that it can be inherited

guillempages committed:
Reduce calls to tankerkoenig api

Use a master/slave concept for sensors; one master gets the data and
updates it into the slaves.

guillempages committed:
Update requirements files

r15ch13 commented:
Just some thoughts:

Is it necessary to add a dependency to implement two simple web requests?

`getPriceList()` (`prices.php`) should be called with a list of station ids instead of one single Id to reduce the load on the API (too many updates will likely get one banned).
You could update all stations/sensors with one single request.

It would be nice to add stations manually. (e.g. to show only stations on a commute route)

guillempages committed:
Update all gas stations at once

guillempages committed:
Remove tests directory

Currently there are no tests for the integration; will be added in
a future commit.

guillempages commented:
> 
> 
> Just some thoughts:
> 
> Is it necessary to add a dependency to implement two simple web requests?

I don't remember where I read that, but I was trying to follow the guideline "do not do HTTP parsing in the integration itself but use a pip library". Maybe I misread that, but since it works, I'm not very keen on changing this now.

> 
> `getPriceList()` (`prices.php`) should be called with a list of station ids instead of one single Id to reduce the load on the API (too many updates will likely get one banned).
> You could update all stations/sensors with one single request.

Well spotted. I was actually trying to do this (that's why there is the whole master/slave construct there), but somehow the code doing many requests at once got lost in some git stash :-(
Fixed now. Thanks!

> 
> It would be nice to add stations manually. (e.g. to show only stations on a commute route)
This PR is just the initial code, so that the integration can be used.
I was planning to allow using entities for getting the location, and did not think about putting extra station ids directly. It is a good idea, but I'm not sure when I can implement that. I'd like to postpone that to a future update.

guillempages committed:
Fix slaves not being updated

Let the base class regularly poll, so that slaves are also updated

guillempages committed:
Refactor entity creation

Create an auxiliary method to add a station to the entity list,
in preparation to allowing extra stations.

guillempages committed:
Add possibility to manually add stations

Add a new configuration option "stations" to manually add extra stations

guillempages commented:
The requirement for the 3rd party lib was from the [development checklist](https://developers.home-assistant.io/docs/en/development_checklist.html) page, so I tried following that. If I misunderstood that, I'd have to do the HTTP parsing in the integration code, but I personally don't think it is worth the effort.

Now, there are some requests being made to the server on initialization (one for the stations within the search radius, and one for each extra station manually added). The updates are done at once (i.e. with a single call) for all stations and fuel types; this should minimize the number of accesses to the server.

There is the possibility to (in addition to the stations found in the search radius) manually add stations; with a "stations:" list. Also updated the documentation PR to reflect that.

basnijholt commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,170 @@
+            self._slaves[station_id].append(slave)
+        self._monitored_stations = list(set(self._monitored_stations) | {station_id})
+
+    async def async_update(self):
```
Why is this `async_update`? Nothing is `await`ed.

guillempages replied:
The example I based my code on was using async calls, and thus I kept everything async. Is this wasting resources, or just a matter of taste?

frenck replied:
You have to change this. Mixing async/sync may cause blocking of our async loop, thus the whole system is affected.

A general rule of thumb:

- If library is async -> integration is async
- If library is sync -> integration is sync

basnijholt commented on homeassistant/components/tankerkoenig/base_sensor.py:
```diff
@@ -0,0 +1,88 @@
+ATTR_STATION_NAME = "station_name"
+ATTRIBUTION = "Data provided by https://creativecommons.tankerkoenig.de"
+
+ICON = "mdi:fuel"
```
I don't think you need a global for this (I could be wrong), just return it in the property.

frenck replied:
That is a choice, nothing wrong with this method.

guillempages replied:
I copied this part from the Prezzibenzina integration, and it is also doing it this way. In general, I like having the constants as globals and grouped, to make it easier to find and change them if needed. If this is blocking the merge, I would change that, but otherwise, I'd prefer to keep it as is.

basnijholt commented on homeassistant/components/tankerkoenig/base_sensor.py:
```diff
@@ -0,0 +1,88 @@
+        self._name = name
+        self._latitude = station["lat"]
+        self._longitude = station["lng"]
+        if station["isOpen"]:
```
`self._is_open = STATE_OPEN if station["isOpen"] else STATE_CLOSED`

guillempages replied:
Done. I hadn't done python for a while, so I agree that my code is not very "pythonic" ;-)

basnijholt commented on homeassistant/components/tankerkoenig/base_sensor.py:
```diff
@@ -0,0 +1,88 @@
+    def new_data(self, data):
+        """Update the internal sensor data."""
+        self._data = data
+        if self._data["status"] == "open":
```
self._is_open = STATE_OPEN if station["status"] == "open" else STATE_CLOSED

guillempages replied:
if you mean "...if data["status"]..." then done ;-)

basnijholt replied:
Exactly!

basnijholt commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,170 @@
+    _LOGGER.debug(
+        "Add_station called for station: %s and fuel types: %s", station, fuel_types
+    )
+    station_id = station["id"]
```
use this in the `else` clause, it is not needed before.

guillempages replied:
Even better, I don't need the variable at all. Thanks! It was probably a remnant of my debugging code, where I had logs everywhere.

guillempages committed:
Fix style issues

Make the code more pythonic

fabaff commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,169 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_RADIUS = 5
```
From my point of view there is no "good" default for the radius. 5 km is way to large if you are in a city and might be too small if you are based on the country side. The website uses 2 km as default.

guillempages replied:
I've changed the default to match the one from the website. Or do you think it would be better to make the radius a required configuration parameter?

fabaff reviewed (changes requested):
The whole master/slave stuff and `base_sensor.py` has to go because this doesn't follow the common practices to write integrations and I don't see any benefits to introduce this, especially as it will make maintenance harder and to review it in the first place.

Also, it looks like that `pytankerkoenig` is not processing the data. A lot of the evaluation is done in the sensor.

guillempages commented on homeassistant/components/tankerkoenig/base_sensor.py:
```diff
@@ -0,0 +1,82 @@
+ATTR_ADDRESS = "address"
+ATTR_BRAND = "brand"
+ATTR_FUEL = "fuel_type"
+ATTR_IS_OPEN = "state"
```
I am using the "STATE_CLOSED" and "STATE_OPEN" constants for this attribute, so that the attribute is "state: open" or "state: closed".
If I change this to "is_open", it sounds strange: "is_open: open" or "is_open: closed". Or do you mean I should have just a boolean? "is_open: True"?

guillempages replied:
I renamed the attribute to "is_open" as requested

guillempages commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,169 @@
+        vol.Optional(CONF_RADIUS, DEFAULT_RADIUS): vol.All(
+            cv.positive_int, vol.Range(min=1)
+        ),
+        vol.Optional(CONF_SCAN_INTERVAL, SCAN_INTERVAL): cv.time_period,
```
Is this available by default on all entities?

frenck replied:
It is part of the `PLATFORM_SCHEMA` you are extending.

See:

https://github.com/home-assistant/home-assistant/blob/3dd5222ec7776dd4771ee9151b3de2d5ab7ff2b6/homeassistant/helpers/config_validation.py#L708-L714

guillempages replied:
OK, I had overlooked that. Thanks!

guillempages committed:
Remove redundant code

Implement suggestions from the code review

guillempages commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,169 @@
+                fuel_types,
+                entities,
+                master,
+            )
```
This is already checked in line 68; it is the same api_key; if it worked once, it should also work here.

The loop is going through the list of extra stations requested by the user; printing a warning but ignoring stations with wrong station ID.

Do you mean, if any station is not available (because of typo), the whole setup should fail, instead of showing the stations that could be found?

guillempages commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,169 @@
+    @property
+    def should_poll(self):
+        """Poll regularly for the standalone sensor."""
+        return True
```
Thanks. This was a leftover from some tests I was doing with the master/slave concept

guillempages commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,169 @@
+    STATE_OPEN,
+)
+from .const import NAME, CONF_STATIONS, CONF_TYPES, FUEL_TYPES
+from .base_sensor import FuelPriceSensorBase
```
Oops. sorry. I tried that manually, but isort is really easier. Thanks

guillempages commented:
> The whole master/slave stuff and `base_sensor.py` has to go because this doesn't follow the common practices to write integrations and I don't see any benefits to introduce this, especially as it will make maintenance harder and to review it in the first place.
> 
The master/slave concept is the solution I found, to avoid hammering the server with requests that could be grouped together.
Actually I tried setting up a platform taking care of getting the data, and have some dummy sensors created from there, but somehow I did not succeed :-( That's why I came up with that concept.

I'll try revisiting the platform concept again. Thanks for the exhaustive review; that was very thorough!

> Also, it looks like that `pytankerkoenig` is not processing the data. A lot of the evaluation is done in the sensor.
pytankerkoenig is just fetching the data; I am not the owner of that library; I just found it and used it. There is actually not much processing to do; just fetching the data from the dictionary. I'm not sure what could be moved into the library additionally.

guillempages commented on homeassistant/components/tankerkoenig/base_sensor.py:
```diff
@@ -0,0 +1,82 @@
+
+ATTR_ADDRESS = "address"
+ATTR_BRAND = "brand"
+ATTR_FUEL = "fuel_type"
```
done

guillempages commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,169 @@
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set the Tankerkoenig sensor platform up."""
+    fuel_types = config.get(CONF_TYPES, FUEL_TYPES)
```
done

guillempages commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,169 @@
+    latitude = config.get(CONF_LATITUDE, hass.config.latitude)
+    longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
+    radius = config.get(CONF_RADIUS, DEFAULT_RADIUS)
+    additional_stations = config.get(CONF_STATIONS, [])
```
done

guillempages commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,169 @@
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_TYPES, ["e5", "e10", "diesel"]): vol.All(
```
done

guillempages commented on homeassistant/components/tankerkoenig/const.py:
```diff
@@ -0,0 +1,9 @@
+DOMAIN = "tankerkoenig"
+NAME = "tankerkoenig"
+
+CONF_TYPES = "fuel_types"
```
done

guillempages committed:
Change to platform component

Remove the master/slave concept, in favor of a platform with dummy sensors.
The platform takes care of contacting the server and fetching updates
atomically, and updating the data on the sensors.

guillempages commented:
I've rewritten the component as a platform component now, and removed the master/slave concept.

It feels cleaner now, thanks for the feedback!

pvizeli commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,100 @@
+            ATTR_ADDRESS: self._address,
+            ATTR_LATITUDE: self._latitude,
+            ATTR_LONGITUDE: self._longitude,
+            ATTR_STATE: self._open_state,
```
`ATTR_IS_OPEN: bool`

A state attribute inside attributes is not possible. If you want monitor the open/close, you can create binary sensors.

guillempages replied:
What about the "flunearyou", "iqvia" and "nut" integrations? But anyway, should I rename it to "status" instead? Or maybe better "current state"?

guillempages replied:
I renamed the attribute to is_open, as requested.

guillempages committed:
Rename ATTR_STATE

Rename the attribute to "IS_OPEN", to avoid confusion with the sensor state.

springstan commented:
Any updates on this PR? :) Would love to have this in HA 🎉

guillempages commented:
> 
> 
> Any updates on this PR? :) Would love to have this in HA 🎉

I've done the requested changes, I'm just waiting for @fabaff to re-review. I also would like to have this in ;-p

springstan commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,100 @@
+        self._latitude = station["lat"]
+        self._longitude = station["lng"]
+        self._is_open = STATE_OPEN if station["isOpen"] else STATE_CLOSED
+        self._address = f"{station['street']} {station['houseNumber']}, {station['postCode']} {station['place']}"
```
Instead of concatenating a string of the address use a dictionary.

guillempages replied:
What do you mean? Should I have one attribute for each part (i.e. one attribute "street", one attribute "houseNumber", ...)?

springstan replied:
No, I was thinking about something along the lines of:
```python
self._address = {
    "street": station["street"]
    "house_number":station["houseNumber"]
    "post_code": station["postCode"]
    "place": station["place"]
}
```

guillempages replied:
But then I need to either concatenate the string anyway in the device_state_attributes method when filling the attributes, or create one attribute anyway for each address sub-field. Or is it possible to have a "dictionary" attribute? I don't remember seeing that on any of the integrations that I am using, but of course I haven't looked into all of them.

I'll create 4 different attributes for the different parts of the address, if it is ok with you.

springstan replied:
> I'll create 4 different attributes for the different parts of the address, if it is ok with you.

Yeah that is okay as well.

springstan commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,100 @@
+        }
+        return attrs
+
+    def new_data(self, data):
```
Why do you create a new method new_data() instead of using update()?

guillempages replied:
AFAIK, update does not take arguments, and I need to pass the new data to the sensor

springstan replied:
Yes, you are completely right.

springstan commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,100 @@
+        else:
+            self._is_open = STATE_OPEN if data["status"] == "open" else STATE_CLOSED
+            self._price = data.get(self._fuel_type)
+        self.update()
```
Why is this called? What functionality or purposes does it have?

guillempages replied:
The platform gets the new fuel prices for all stations at once, and distributes them to the sensors (one sensor for each combination station/fuel type). Afterwards I have to call the "update" method, so that the sensor's value gets updated in the system.

springstan replied:
Okay thanks for the clarification.

guillempages committed:
Minor updates

Combine two consecutive error logs into a single one.
Update the sensor's icon

guillempages committed:
Separate address into different fields

springstan commented on homeassistant/components/tankerkoenig/__init__.py:
```diff
@@ -0,0 +1,203 @@
+
+    def __init__(self, hass, conf):
+        """Initialize the data object."""
+        self._api_key = conf.get(CONF_API_KEY)
```
Use conf[CONF_API_KEY] here instead, since this parameter is required.

springstan commented on homeassistant/components/tankerkoenig/__init__.py:
```diff
@@ -0,0 +1,203 @@
+                )
+                self._entities[station_id].append(station_sensor)
+            else:
+                _LOGGER.info("Station %s does not offer %s fuel", station_id, fuel)
```
Use a warning here instead.

guillempages committed:
Style updates

Use "[]" syntax instead of ".get()" for required parameters
Use warning log level for not available fuel types

springstan commented on homeassistant/components/tankerkoenig/__init__.py:
```diff
@@ -0,0 +1,203 @@
+    def update(self, now=None):
+        """Get the latest data from tankerkoenig.de."""
+        _LOGGER.debug("Fetching new data from tankerkoenig.de")
+        data = pytankerkoenig.getPriceList(self._api_key, list(self._entities.keys()))
```
Extract `list(self._entities.keys()` into a variable before and use it here to make it more readable.

springstan commented on homeassistant/components/tankerkoenig/manifest.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "domain": "tankerkoenig",
+  "name": "Tankerkoenig",
+  "config_flow": false,
```
Remove this line since `false` is the default value.

guillempages committed:
Implement review comments

Fix style issues
Improve error messages
Remove redundant options

springstan commented on homeassistant/components/tankerkoenig/__init__.py:
```diff
@@ -0,0 +1,204 @@
+
+    conf = config[DOMAIN]
+
+    _LOGGER.debug("Setting up tankerkoenig")
```
There is no need to add the platform or component name to the log messages. See the [style guidelines](https://developers.home-assistant.io/docs/en/development_guidelines.html#log-messages).

springstan reviewed (commented):
@MartinHjelmare what do you think of this PR?

springstan reviewed (approved):
LGTM ✌

MartinHjelmare commented on homeassistant/components/tankerkoenig/__init__.py:
```diff
@@ -0,0 +1,204 @@
+        _LOGGER.debug("Fetching data for (%s,%s) rad: %s", latitude, longitude, radius)
+
+        try:
+            data = pytankerkoenig.getNearbyStations(
```
We don't want side effects in init method. Please either pass this instance to `TankerkoenigData` or move this to a dedicated setup method.

MartinHjelmare commented on homeassistant/components/tankerkoenig/__init__.py:
```diff
@@ -0,0 +1,204 @@
+        # Add stations found via location + radius
+        nearby_stations = data["stations"]
+        if not nearby_stations:
+            _LOGGER.warning(
```
Please move side effects like logging out of init method.

MartinHjelmare commented on homeassistant/components/tankerkoenig/__init__.py:
```diff
@@ -0,0 +1,204 @@
+        )
+        for fuel in self._fuel_types:
+            if fuel in station.keys():
+                station_sensor = FuelPriceSensor(
```
The entity is a platform concern and should be contained within its platform. The entity should be created and added within its platform.

MartinHjelmare commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,109 @@
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the tankerkoenig sensors."""
+    tankerkoenig = hass.data[DOMAIN]
```
Please add a guard clause here that checks if discovery_info is None and return if so. We need to guard against the user adding a platform config section by mistake which would load the platform twice.

MartinHjelmare commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,109 @@
+    """Set up the tankerkoenig sensors."""
+    tankerkoenig = hass.data[DOMAIN]
+    _LOGGER.debug("Setup platform. Stations: %s ", tankerkoenig.entity_list)
+    async_add_entities(tankerkoenig.entity_list)
```
Create the entities here. We can use the tankerkoenig api instance to fetch relevant info.

MartinHjelmare commented on homeassistant/components/tankerkoenig/__init__.py:
```diff
@@ -0,0 +1,204 @@
+            _LOGGER.debug("Received data: %s", data)
+            for station_id, station_list in self._entities.items():
+                for station in station_list:
+                    station.new_data(data["prices"].get(station_id))
```
See https://developers.home-assistant.io/docs/en/integration_fetching_data.html for our recommended way to fetch and update data.

This looks like there's a single api that provides all entity data, so we should use the `DataUpdateCoordinator`.
https://developers.home-assistant.io/docs/en/integration_fetching_data.html#coordinated-single-api-poll-for-data-for-all-entities

guillempages replied:
This PR was done before the coordinated data updater was introduced into home assistant, so I had no chance of using this when I created the commits.

Nevertheless, I rebased my commits and tried implementing it as requested, to the best of my abilities.

By doing that, I found an inconsistency in the documentation you pointed to:
In the example, the ```available``` property is implemented as ```return not self.coordinator.failed_last_update```, but should be ```return self.coordinator.last_update_success``` instead (I got an error otherwise).

MartinHjelmare reviewed (changes requested):
The integration needs to be refactored to follow our design rules.

guillempages committed:
Refactor using DataUpdateCoordinator

Use the new DataUpdateCoordinator to fetch the global data from
the API, instead of implementing an own method.
Also fix comments from the PR

MartinHjelmare commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,160 @@
+                fuel, station, coordinator, f"{NAME}_{station['name']}_{fuel}"
+            )
+            _LOGGER.debug("Added sensor %s", sensor.name)
+            entity_list.append(sensor)
```
```suggestion
            entities.append(sensor)
```

MartinHjelmare commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,160 @@
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+    async_add_entities(entity_list)
```
```suggestion
    async_add_entities(entities)
```

MartinHjelmare commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,160 @@
+class FuelPriceSensor(Entity):
+    """Contains prices for fuel in a given station."""
+
+    def __init__(self, fuel_type, station, coordinator, name=NAME):
```
```suggestion
    def __init__(self, fuel_type, station, coordinator, name):
```

MartinHjelmare commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,160 @@
+    def device_state_attributes(self):
+        """Return the attributes of the device."""
+        data = self._coordinator.data[self._station_id]
+        if data is None or "status" not in data.keys():
```
```suggestion
        if data is None or "status" not in data:
```

MartinHjelmare commented on homeassistant/components/tankerkoenig/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        """Return the attributes of the device."""
+        data = self._coordinator.data[self._station_id]
+        if data is None or "status" not in data.keys():
+            self._is_open = STATE_UNKNOWN
```
Our standard is to not add the item at all to state attributes if the value is unknown.

guillempages committed:
Implement PR comments

Implement suggestions to improve code readability and keep the Home Assistant style.
Also separate fetching data to an async thread

MartinHjelmare reviewed (approved):
Nice!

guillempages commented:
> 
> 
> Nice!

Thanks for the thorough and detailed review; it was very easy to follow and implement.

@fabaff , can you check as well, and see if we can get this in?

guillempages commented:
@fabaff is there still something you would like me to change in the PR? It has been reworked a lot since your initial comments

MartinHjelmare commented:
Let's merge. There's time before next release if there are more comments.
